File: clients/node/src/framework/engine.js
'use strict';

const { MsgType } = require('./net/protocol');
const { loadGameClasses } = require('./lobby/session');
const { BOLD, RESET, DIM, GREEN } = require('./ui/terminal');
const { renderGameStr } = require('./ui/lobby_screen');
const { t } = require('./i18n');

class GameEngine {
  constructor({ name, chatLog, panel }) {
    this.name = name;
    this.chatLog = chatLog;
    this.panel = panel;
    this.sendMove = null;
    this.sendChat = null;
    this.gameObj = null;
    this.players = [];
  }

  onMessage(msg) {
    const type = msg.type;
    if (type === MsgType.CHAT) {
      this.chatLog.add(msg.from || '?', msg.text || '');
      this.panel.pushChat(msg.from || '?', msg.text || '');
    } else if (type === MsgType.PLAYER_LIST) {
      const newPlayers = msg.players || [];
      if (this.gameObj && !this.gameObj._over && this.players.length > newPlayers.length) {
        const leaver = this.players.find(p => !newPlayers.includes(p));
        if (leaver) {
          this.gameObj._over = true;
          this.gameObj._winner = newPlayers[0] || null;
          this.gameObj._forfeitedBy = leaver;
        }
      }
      this.players = newPlayers;
    } else if (type === MsgType.GAME_START) {
      const classes = loadGameClasses();
      const gname = msg.game;
      if (classes[gname]) {
        this.gameObj = new classes[gname]();
        this.gameObj.start(msg.players || []);
      }
    } else if (type === MsgType.STATE) {
      if (this.gameObj && typeof this.gameObj.loadState === 'function') {
        this.gameObj.loadState(msg.data || {}, this.name);
        this.panel.setContent(renderGameStr(this.gameObj, this.players, this.chatLog.recent(3), this.name));
      }
    } else if (type === MsgType.GAME_OVER) {
      if (this.gameObj) {
        this.gameObj._over = true;
        this.gameObj._winner = msg.winner || null;
      }
    }
  }

  async run() {
    while (true) {
      const gameObj = this.gameObj;

      if (!gameObj) {
        this.panel.setContent('\n  Waiting for players...');
        const k = await this.panel.waitKey(500);
        if (k === 'q') { this.panel.destroy(); return; }
        continue;
      }

      this.panel.setContent(renderGameStr(gameObj, this.players, this.chatLog.recent(3), this.name));

      const [done, winner] = gameObj.isOver();
      if (done) {
        const endMsg = gameObj._forfeitedBy
          ? `${gameObj._forfeitedBy} disconnected. ${winner ? winner + ' wins by forfeit!' : ''}`
          : winner ? t('game.winner', { winner }) : t('game.draw');
        this.panel.setContent(`\n  ${BOLD}Game Over${RESET}\n\n  ${endMsg}\n\n`);
        await this.panel.readInput('  Press Enter to exit... ', true);
        this.panel.destroy();
        return;
      }

      if (gameObj.currentTurn() === this.name) {
        const raw = await this.panel.readInput(t('game.move_prompt'));
        const cmd = raw.trim().toLowerCase();
        if (!cmd) continue;
        if (cmd === 'q') { this.panel.destroy(); return; }
        if (cmd === '?') {
          const lines = ['', `  ${BOLD}? Help${RESET}`, '', ...gameObj.getHelp().map(l => `  ${l}`), '',
            `  ${DIM}c = send chat   t = toggle chat panel   ? = help   q = quit${RESET}`, ''];
          this.panel.setContent(lines.join('\n'));
          await this.panel.readInput('  Press Enter to return... ', true);
          continue;
        }
        if (cmd === 'c') {
          const msg = await this.panel.readInput(t('game.chat_prompt'), true);
          if (msg.trim() && this.sendChat) {
            this.sendChat(msg.trim());
            await new Promise(r => setTimeout(r, 60));
          }
          continue;
        }
        const parsed = gameObj.parseInput(raw.trim());
        if (!parsed) {
          this.panel.setPrompt(`  ${DIM}Unrecognized input — type ? for help${RESET}`);
          await new Promise(r => setTimeout(r, 1200));
          this.panel.setPrompt('');
          continue;
        }
        if (this.sendMove) this.sendMove(parsed);
      } else {
        const k = await this.panel.waitKey(150);
        if (k === 'q') { this.panel.destroy(); return; }
        if (k === 'c') {
          const msg = await this.panel.readInput(t('game.chat_prompt'), true);
          if (msg.trim() && this.sendChat) {
            this.sendChat(msg.trim());
            await new Promise(r => setTimeout(r, 60));
          }
        } else if (k === '?') {
          const lines = ['', `  ${BOLD}? Help${RESET}`, '', ...gameObj.getHelp().map(l => `  ${l}`), '',
            `  ${DIM}c = send chat   t = toggle chat panel   ? = help   q = quit${RESET}`, ''];
          this.panel.setContent(lines.join('\n'));
          await this.panel.readInput('  Press Enter to return... ', true);
        }
      }
    }
  }
}

module.exports = { GameEngine };

File: clients/python/framework/engine.py
import time
import sys
import json

from framework.net.protocol import MsgType
from framework.lobby.session import _load_game_classes
from framework.ui.terminal import clear, header, read_line_with_refresh, DIM, BOLD, RESET
from framework.ui.lobby_screen import render_game
from framework.i18n import t


class GameEngine:
    """Owns the game loop: turn management, input dispatch, state sync, chat."""

    def __init__(self, *, name: str, chat_log):
        self.name = name
        self.chat_log = chat_log
        self.send_move = None   # set by caller after construction
        self.send_chat = None   # set by caller after construction
        self.game_obj = None
        self.players: list = []
        self._last_snap = None

    # ── network callback (called from TCP receive thread) ──────────────────

    def on_message(self, msg: dict) -> None:
        msg_type = msg.get('type')
        if msg_type == MsgType.CHAT:
            self.chat_log.add(msg.get('from', '?'), msg.get('text', ''))
        elif msg_type == MsgType.PLAYER_LIST:
            new_players = list(msg.get('players', []))
            if (self.game_obj is not None
                    and not getattr(self.game_obj, '_over', False)
                    and len(self.players) > len(new_players)):
                leaver = next((p for p in self.players if p not in new_players), None)
                if leaver:
                    self.game_obj._over = True
                    self.game_obj._winner = new_players[0] if new_players else None
                    self.game_obj._forfeited_by = leaver
            self.players = new_players
        elif msg_type == MsgType.GAME_START:
            classes = _load_game_classes()
            game_name = msg.get('game')
            if game_name in classes:
                self.game_obj = classes[game_name]()
                self.game_obj.start(msg.get('players', []))
        elif msg_type == MsgType.STATE:
            if self.game_obj is not None:
                self.game_obj.load_state(msg.get('data', {}), self.name)
        elif msg_type == MsgType.GAME_OVER:
            if self.game_obj is not None:
                self.game_obj._over = True
                self.game_obj._winner = msg.get('winner')

    # ── main loop (runs on main thread) ────────────────────────────────────

    def run(self) -> None:
        self._last_snap = None
        while True:
            game = self.game_obj
            if game is None:
                s = self._snap()
                if s != self._last_snap:
                    clear()
                    header(t('game.waiting'))
                    self._last_snap = s
                time.sleep(0.5)
                continue

            s = self._snap()
            if s != self._last_snap:
                clear()
                render_game(game, self.players, self.chat_log.recent(3), self.name)
                self._last_snap = s

            done, winner = game.is_over()
            if done:
                clear()
                header(t('game.over'))
                forfeited_by = getattr(game, '_forfeited_by', None)
                if forfeited_by:
                    end_msg = (f'{forfeited_by} disconnected. '
                               f'{winner + " wins by forfeit!" if winner else ""}')
                else:
                    end_msg = t('game.winner', winner=winner) if winner else t('game.draw')
                print(f'\n  {BOLD}{end_msg}{RESET}\n')
                input('  ' + t('game.continue'))
                return

            if game.current_turn() == self.name:
                def _snap_check():
                    s = self._snap()
                    if s != self._last_snap:
                        clear()
                        render_game(game, self.players, self.chat_log.recent(3), self.name)
                        self._last_snap = s
                        return True
                    return False

                raw = read_line_with_refresh(t('game.move_prompt'), on_check=_snap_check).strip()
                if not raw:
                    self._last_snap = None
                    continue
                cmd = raw.lower()
                if cmd == 'q':
                    return
                if cmd == 't':
                    msg = input(t('game.chat_prompt')).strip()
                    if msg and self.send_chat:
                        self.send_chat(msg)
                        time.sleep(0.06)
                    self._last_snap = None
                    continue
                if cmd == '?':
                    self._show_help(game)
                    input('  ' + t('game.continue'))
                    self._last_snap = None
                    continue
                parsed = game.parse_input(raw)
                if parsed is None:
                    print(f'\n  {DIM}Unrecognized input — type ? for help{RESET}\n')
                    time.sleep(1.2)
                    self._last_snap = None
                    continue
                if self.send_move:
                    self.send_move(parsed)
                self._last_snap = None
            else:
                ch = self._peek_char()
                if ch == 'q':
                    return
                if ch == 't':
                    msg = input(t('game.chat_prompt')).strip()
                    if msg and self.send_chat:
                        self.send_chat(msg)
                        time.sleep(0.06)
                    self._last_snap = None
                elif ch == '?':
                    self._show_help(game)
                    input('  ' + t('game.continue'))
                    self._last_snap = None
                else:
                    time.sleep(0.15)

    # ── helpers ─────────────────────────────────────────────────────────────

    def _snap(self) -> str:
        g = self.game_obj
        if g is None:
            return '__waiting__'
        chats = '|'.join(
            f"{e['from']}:{e['text']}" for e in self.chat_log.recent(3)
        )
        return json.dumps(g.get_state(self.name), default=str) + '|' + chats

    def _show_help(self, game) -> None:
        clear()
        header('? Help')
        for line in game.get_help():
            print(f'  {line}')
        print()
        print(f'  {DIM}t = chat   ? = this help{RESET}')
        print()

    @staticmethod
    def _peek_char() -> str | None:
        if sys.platform == 'win32':
            import msvcrt
            if msvcrt.kbhit():
                return msvcrt.getwch().lower()
        return None
